{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 7 - RNA-seq and bread"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!hisat2-build ref.fasta data/gen_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!hisat2 -p 4 -x data/gen_index -U SRR941816/SRR941816.fastq | samtools sort > aln_sorted_0min_1_out.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "!gffread annotation_genomic.gff -T -o annotation_genomic.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
      "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
      "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
      "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
      "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
      "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
      "\t  v2.0.0\n",
      "\n",
      "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Input files : \u001b[36m4 BAM files  \u001b[0m \u001b[0m                                   ||\n",
      "||                           \u001b[32mo\u001b[36m aln_sorted_0min_1_out.bam\u001b[0m \u001b[0m                     ||\n",
      "||                           \u001b[32mo\u001b[36m aln_sorted_0min_2_out.bam\u001b[0m \u001b[0m                     ||\n",
      "||                           \u001b[32mo\u001b[36m aln_sorted_30min_1_out.bam\u001b[0m \u001b[0m                    ||\n",
      "||                           \u001b[32mo\u001b[36m aln_sorted_30min_2_out.bam\u001b[0m \u001b[0m                    ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||             Output file : \u001b[36mfeature_counts.txt\u001b[0m \u001b[0m                              ||\n",
      "||                 Summary : \u001b[36mfeature_counts.txt.summary\u001b[0m \u001b[0m                      ||\n",
      "||              Annotation : \u001b[36mannotation_genomic.gtf (GTF)\u001b[0m \u001b[0m                    ||\n",
      "||      Dir for temp files : \u001b[36m./\u001b[0m \u001b[0m                                              ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
      "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
      "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
      "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n",
      "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Load annotation file annotation_genomic.gtf \u001b[0m... \u001b[0m                           ||\n",
      "||    Features : \u001b[36m6760\u001b[0m \u001b[0m                                                        ||\n",
      "||    Meta-features : \u001b[36m6420\u001b[0m \u001b[0m                                                   ||\n",
      "||    Chromosomes/contigs : \u001b[36m17\u001b[0m \u001b[0m                                               ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file aln_sorted_0min_1_out.bam... \u001b[0m                             ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m9749632\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m7291722 (74.8%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.14 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file aln_sorted_0min_2_out.bam... \u001b[0m                             ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m10810498\u001b[0m \u001b[0m                                            ||\n",
      "||    Successfully assigned alignments : \u001b[36m7987002 (73.9%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.16 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file aln_sorted_30min_1_out.bam... \u001b[0m                            ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m1880817\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m1402166 (74.6%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.03 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Process BAM file aln_sorted_30min_2_out.bam... \u001b[0m                            ||\n",
      "||    Single-end reads are included. \u001b[0m                                         ||\n",
      "||    Total alignments : \u001b[36m6782905\u001b[0m \u001b[0m                                             ||\n",
      "||    Successfully assigned alignments : \u001b[36m4975463 (73.4%)\u001b[0m \u001b[0m                     ||\n",
      "||    Running time : \u001b[36m0.10 minutes\u001b[0m \u001b[0m                                            ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "|| Summary of counting results can be found in file \"feature_counts.txt.summ \u001b[0m ||\n",
      "|| ary\" \u001b[0m                                                                      ||\n",
      "||  \u001b[0m                                                                          ||\n",
      "\\\\============================================================================//\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!subread_2/bin/featureCounts -g gene_id -a annotation_genomic.gtf -o feature_counts.txt \"aln_sorted_0min_1_out.bam\" \"aln_sorted_0min_2_out.bam\" \"aln_sorted_30min_1_out.bam\" \"aln_sorted_30min_2_out.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat feature_counts.txt | cut -f 1,7-10 > simple_counts.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 3.6.2 (2019-12-12) -- \"Dark and Stormy Night\"\n",
      "Copyright (C) 2019 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-apple-darwin15.6.0 (64-bit)\n",
      "\n",
      "R -- это свободное ПО, и оно поставляется безо всяких гарантий.\n",
      "Вы вольны распространять его при соблюдении некоторых условий.\n",
      "Введите 'license()' для получения более подробной информации.\n",
      "\n",
      "R -- это проект, в котором сотрудничает множество разработчиков.\n",
      "Введите 'contributors()' для получения дополнительной информации и\n",
      "'citation()' для ознакомления с правилами упоминания R и его пакетов\n",
      "в публикациях.\n",
      "\n",
      "Введите 'demo()' для запуска демонстрационных программ, 'help()' -- для\n",
      "получения справки, 'help.start()' -- для доступа к справке через браузер.\n",
      "Введите 'q()', чтобы выйти из R.\n",
      "\n",
      "> # Load the library.\n",
      "> library(DESeq2)\n",
      "Загрузка требуемого пакета: S4Vectors\n",
      "Загрузка требуемого пакета: stats4\n",
      "Загрузка требуемого пакета: BiocGenerics\n",
      "Загрузка требуемого пакета: parallel\n",
      "\n",
      "Присоединяю пакет: ‘BiocGenerics’\n",
      "\n",
      "Следующие объекты скрыты от ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "Следующие объекты скрыты от ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "Следующие объекты скрыты от ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which, which.max, which.min\n",
      "\n",
      "\n",
      "Присоединяю пакет: ‘S4Vectors’\n",
      "\n",
      "Следующий объект скрыт от ‘package:base’:\n",
      "\n",
      "    expand.grid\n",
      "\n",
      "Загрузка требуемого пакета: IRanges\n",
      "Загрузка требуемого пакета: GenomicRanges\n",
      "Загрузка требуемого пакета: GenomeInfoDb\n",
      "Загрузка требуемого пакета: SummarizedExperiment\n",
      "Загрузка требуемого пакета: Biobase\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "Загрузка требуемого пакета: DelayedArray\n",
      "Загрузка требуемого пакета: matrixStats\n",
      "\n",
      "Присоединяю пакет: ‘matrixStats’\n",
      "\n",
      "Следующие объекты скрыты от ‘package:Biobase’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "Загрузка требуемого пакета: BiocParallel\n",
      "\n",
      "Присоединяю пакет: ‘DelayedArray’\n",
      "\n",
      "Следующие объекты скрыты от ‘package:matrixStats’:\n",
      "\n",
      "    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges\n",
      "\n",
      "Следующие объекты скрыты от ‘package:base’:\n",
      "\n",
      "    aperm, apply, rowsum\n",
      "\n",
      "Предупреждения:\n",
      "1: пакет ‘S4Vectors’ был собран под R версии 3.6.3 \n",
      "2: пакет ‘GenomeInfoDb’ был собран под R версии 3.6.3 \n",
      "3: пакет ‘DelayedArray’ был собран под R версии 3.6.3 \n",
      "> \n",
      "> # Set up the conditions based on the experimental setup.\n",
      "> cond_1 = rep(\"cond1\", 2)\n",
      "> cond_2 = rep(\"cond2\", 2)\n",
      "> \n",
      "> # Read the data from the standard input.\n",
      "> countData = read.table(\"stdin\", header=TRUE, sep=\"\\t\", row.names=1 )\n",
      "> \n",
      "> # Build the dataframe from the conditions\n",
      "> samples = names(countData)\n",
      "> condition = factor(c(cond_1, cond_2))\n",
      "> colData = data.frame(samples=samples, condition=condition)\n",
      "> \n",
      "> # Create DESEq2 dataset.\n",
      "> dds = DESeqDataSetFromMatrix(countData=countData, colData=colData, design = ~condition)\n",
      "> \n",
      "> #Set the reference to be compared\n",
      "> dds$condition = relevel(dds$condition,\"cond1\")\n",
      "> \n",
      "> # Run deseq\n",
      "> dds = DESeq(dds)\n",
      "estimating size factors\n",
      "estimating dispersions\n",
      "gene-wise dispersion estimates\n",
      "mean-dispersion relationship\n",
      "final dispersion estimates\n",
      "fitting model and testing\n",
      "> \n",
      "> # Format the results.\n",
      "> res = results(dds)\n",
      "> \n",
      "> # Sort the results data frame by the padj and foldChange columns.\n",
      "> sorted = res[with(res, order(padj, -log2FoldChange)), ]\n",
      "> \n",
      "> # Turn it into a dataframe to have proper column names.\n",
      "> sorted.df = data.frame(\"id\"=rownames(sorted),sorted)\n",
      "> \n",
      "> # Write the table out.\n",
      "> write.table(sorted.df, file=\"result.txt\", sep=\"\\t\", col.names=NA, quote=FALSE)\n",
      "> \n",
      "> # Get normalized counts and write this to a file\n",
      "> nc = counts(dds,normalized=TRUE)\n",
      "> \n",
      "> # Turn it into a dataframe to have proper column names.\n",
      "> dt = data.frame(\"id\"=rownames(nc),nc)\n",
      "> \n",
      "> # Save the normalize data matrix.\n",
      "> write.table(dt, file=\"norm-matrix-deseq2.txt\", sep=\"\\t\",  row.name=FALSE, col.names=TRUE,quote=FALSE)\n",
      "> \n",
      "> \n",
      "> \n"
     ]
    }
   ],
   "source": [
    "!cat simple_counts.txt | R -f deseq2.r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "R version 3.6.2 (2019-12-12) -- \"Dark and Stormy Night\"\n",
      "Copyright (C) 2019 The R Foundation for Statistical Computing\n",
      "Platform: x86_64-apple-darwin15.6.0 (64-bit)\n",
      "\n",
      "R -- это свободное ПО, и оно поставляется безо всяких гарантий.\n",
      "Вы вольны распространять его при соблюдении некоторых условий.\n",
      "Введите 'license()' для получения более подробной информации.\n",
      "\n",
      "R -- это проект, в котором сотрудничает множество разработчиков.\n",
      "Введите 'contributors()' для получения дополнительной информации и\n",
      "'citation()' для ознакомления с правилами упоминания R и его пакетов\n",
      "в публикациях.\n",
      "\n",
      "Введите 'demo()' для запуска демонстрационных программ, 'help()' -- для\n",
      "получения справки, 'help.start()' -- для доступа к справке через браузер.\n",
      "Введите 'q()', чтобы выйти из R.\n",
      "\n",
      "> library(gplots)\n",
      "\n",
      "Присоединяю пакет: ‘gplots’\n",
      "\n",
      "Следующий объект скрыт от ‘package:stats’:\n",
      "\n",
      "    lowess\n",
      "\n",
      "> \n",
      "> # Read normalized counts\n",
      "> data = read.table(\"stdin\", header=T, sep=\"\\t\", as.is=TRUE)\n",
      "> \n",
      "> gene = data[,1]\n",
      "> vals = as.matrix(data[,2:ncol(data)])\n",
      "> \n",
      "> # Adds a little noise to each element\n",
      "> # To avoid the clusteing function failing on zero\n",
      "> # variance datalines.\n",
      "> vals = jitter(vals, factor = 1, amount=0.00001)\n",
      "> \n",
      "> \n",
      "> # Calculate zscore\n",
      "> score = NULL\n",
      "> for (i in 1:nrow(vals)) {\n",
      "+     row=vals[i,]\n",
      "+     zscore=(row-mean(row))/sd(row)\n",
      "+     score =rbind(score,zscore)\n",
      "+ }\n",
      "> \n",
      "> row.names(score) = gene\n",
      "> zscore=score\n",
      "> \n",
      "> # Generate heatmap\n",
      "> mat = as.matrix(zscore)\n",
      "> \n",
      "> # Opent the drawing device.\n",
      "> pdf('output.pdf')\n",
      "> \n",
      "> colors = colorRampPalette(c(\"green\",\"black\",\"red\"),space=\"rgb\")(256)\n",
      "> heatmap.2(mat,col=colors,density.info=\"none\",trace=\"none\", margins=c(14,14),lhei=c(1,5))\n",
      "> \n",
      "> invisible(dev.off())\n",
      "> \n"
     ]
    }
   ],
   "source": [
    "!cat norm-matrix-deseq2.txt | R -f draw-heatmap.r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "!head -n 50 result.txt | cut -f 1 | cut -c 6- > genes.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
